util/buildinfo/buildinfo.go

```diff
@@ -2,136 +2,236 @@ package buildinfo
-		return bis[i].Ref < bis[j].Ref
+	sort.Slice(srcs, func(i, j int) bool {
+		return srcs[i].Ref < srcs[j].Ref
 	})
 
-	return json.Marshal(map[string][]exptypes.BuildInfo{
-		"sources": bis,
-	})
+	return srcs, nil
 }
 
-// imageConfigBuildInfo returns build dependencies from image config
-func imageConfigBuildInfo(imageConfig []byte) ([]exptypes.BuildInfo, error) {
-	if len(imageConfig) == 0 {
-		return nil, nil
+// FormatOpts holds build info format options.
+type FormatOpts struct {
+	RemoveAttrs bool
+}
+
+// Format formats build info.
+func Format(dt []byte, format FormatOpts) (_ []byte, err error) {
+	if len(dt) == 0 {
+		return dt, nil
+	}
+	var bi binfotypes.BuildInfo
+	if err := json.Unmarshal(dt, &bi); err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshal buildinfo for formatting")
+	}
+	if format.RemoveAttrs {
+		bi.Attrs = nil
 	}
-	var config struct {
-		BuildInfo []byte `json:"moby.buildkit.buildinfo.v1,omitempty"`
+	if dt, err = json.Marshal(bi); err != nil {
+		return nil, err
+	}
+	return dt, nil
+}
+
+var knownAttrs = []string{
+	//"cmdline",
+	"context",
+	"filename",
+	"source",
+
+	//"add-hosts",
+	//"cgroup-parent",
+	//"force-network-mode",
+	//"hostname",
+	//"image-resolve-mode",
+	//"platform",
+	"shm-size",
+	"target",
+	"ulimit",
+}
+
+// filterAttrs filters frontent opt by picking only those that
+// could effectively change the build result.
+func filterAttrs(attrs map[string]string) map[string]string {
+	filtered := make(map[string]string)
+	for k, v := range attrs {
+		// Control args are filtered out
+		if isControlArg(k) {
+			continue
+		}
+		// Always include args and labels
+		if strings.HasPrefix(k, "build-arg:") || strings.HasPrefix(k, "label:") {
+			filtered[k] = v
+			continue
+		}
+		// Filter only for known attributes
+		for _, knownAttr := range knownAttrs {
+			if knownAttr == k {
+				filtered[k] = v
+				break
+			}
+		}
 	}
-	if err := json.Unmarshal(imageConfig, &config); err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal buildinfo from config")
+	return filtered
+}
+
+var knownControlArgs = []string{
+	"BUILDKIT_CACHE_MOUNT_NS",
+	"BUILDKIT_CONTEXT_KEEP_GIT_DIR",
+	"BUILDKIT_INLINE_BUILDINFO_ATTRS",
+	"BUILDKIT_INLINE_CACHE",
+	"BUILDKIT_MULTI_PLATFORM",
+	"BUILDKIT_SANDBOX_HOSTNAME",
+	"BUILDKIT_SYNTAX",
+}
+
+// isControlArg checks if a build attributes is a control arg
+func isControlArg(attrKey string) bool {
+	for _, k := range knownControlArgs {
+		if strings.HasPrefix(attrKey, "build-arg:"+k) {
+			return true
+		}
 	}
-	if len(config.BuildInfo) == 0 {
-		return nil, nil
+	return false
+}
+
+// ReduceMap joins map string (union)
+func ReduceMap(m1, m2 map[string]string) map[string]string {
+	if m2 == nil {
+		return m1
 	}
-	var bi []exptypes.BuildInfo
-	if err := json.Unmarshal(config.BuildInfo, &bi); err != nil {
-		return nil, errors.Wrapf(err, "failed to unmarshal %s", ImageConfigField)
+	for k1, v1 := range m1 {
+		if v2, ok := m2[k1]; ok {
+			m2[k1] = v2
```

Comment:
This line doesn't seem to do anything.

I guess the logic for this merge should be:
- m1 exists, m2 exists -> m1 (keep the frontend value)
- m1 exists, m2 not -> m1
- m1 not, m2 exists -> m2
- m1 nil, m2 exists -> remove value

Reply:
done and add a small test for the attrs merge